libraries/chain/include/eosio/chain/wasm_interface_private.hpp

```diff
@@ -96,13 +101,99 @@ struct eosvmoc_tier {
 #ifdef EOSIO_EOS_VM_OC_RUNTIME_ENABLED
          if(eosvmoc_tierup != wasm_interface::vm_oc_enable::oc_none) {
             EOS_ASSERT(vm != wasm_interface::vm_type::eos_vm_oc, wasm_exception, "You can't use EOS VM OC as the base runtime when tier up is activated");
-            eosvmoc = std::make_unique<eosvmoc_tier>(data_dir, eosvmoc_config, d);
+            eosvmoc = std::make_unique<eosvmoc_tier>(data_dir, eosvmoc_config, d, [this](boost::asio::io_context& ctx, const digest_type& code_id, fc::time_point queued_time) {
+               async_compile_complete(ctx, code_id, queued_time);
+            });
          }
 #endif
       }
 
       ~wasm_interface_impl() = default;
 
+#ifdef EOSIO_EOS_VM_OC_RUNTIME_ENABLED
+      // called from async thread
+      void async_compile_complete(boost::asio::io_context& ctx, const digest_type& code_id, fc::time_point queued_time) {
+         if (executing_code_hash.load() == code_id) { // is action still executing?
+            auto elapsed = fc::time_point::now() - queued_time;
+            ilog("EOS VM OC tier up for ${id} compile complete ${t}ms", ("id", code_id)("t", elapsed.count()/1000));
+            auto expire_in = std::max(fc::microseconds(0), fc::milliseconds(500) - elapsed);
+            std::shared_ptr<boost::asio::steady_timer> timer = std::make_shared<boost::asio::steady_timer>(ctx);
```

Comment:
No reason. I would actually be in favor of immediately interrupting. The issue called for only interrupting if running longer than 500ms.

Reply:
I guess I see my confusion --  "running longer than 500ms" : is that really what is going on here? `queued_time` is when the compilation was started, not how long the transaction or action has been running.

Reply:
True. Seems to me we could just always interrupt and make this a bit simpler.

Reply:
I think maybe it's a good thing we _don't_ always interrupt, otherwise maybe in a worst case pathological situation the block could take 2x the time to apply (it would seem very very hard to coax this to occur but maybe I haven't really thought it through enough)